File: LeetCode/00623. Add One Row to Tree/00623.c
struct TreeNode * addOneRow(struct TreeNode *root, int val, int depth) {
    if (depth == 1) {
        struct TreeNode *newNode = (struct TreeNode *) malloc(sizeof(struct TreeNode));
        *newNode = (struct TreeNode){ .left = root, .right = NULL, .val = val };
        return newNode;
    } else if (depth == 2) {
        struct TreeNode *nodes = (struct TreeNode *) malloc(sizeof(struct TreeNode) * 2);
        *(nodes)    = (struct TreeNode){ .left = root->left,    .right = NULL,          .val = val          };
        *(nodes+1)  = (struct TreeNode){ .left = NULL,          .right = root->right,   .val = val          };
        *root       = (struct TreeNode){ .left = nodes,         .right = nodes+1,       .val = root->val    };
        return root;
    }

    if (root->left != NULL) {
        addOneRow(root->left, val, depth-1);
    }
    if (root->right != NULL) {
        addOneRow(root->right, val, depth-1);
    }

    return root;
}

File: LeetCode/00623. Add One Row to Tree/00623.go
// https://leetcode.com/problems/add-one-row-to-tree

func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
    if depth > 1 {
        if root == nil {
            return nil
        }

        root.Left = addOneRow(root.Left, val, depth-1)        
        root.Right = addOneRow(root.Right, val, depth-1)

        if depth == 2 {
            if root.Right != nil {
                root.Right.Right = root.Right.Left
                root.Right.Left = nil
            }
        }

        return root
    }

    node := &TreeNode{
        Val: val,
        Left: root,
    }

    return node
}
